src/utils/roll-orb.ts

```diff
@@ -0,0 +1,89 @@
+import * as debug from 'debug';
+import * as yaml from 'yaml';
+
+import { MAIN_BRANCH, ORB_KEY, OrbTarget } from '../constants';
+import { getOctokit } from './octokit';
+import { getOrbPRText } from './pr-text-orb';
+
+// Rolls an orb in a .circleci/config.yml file to a new version
+export async function rollOrb({
+  orbTarget: rollTarget,
+  sha,
+  targetValue,
+  repository,
+}): Promise<any> {
+  const d = debug(`roller/orb/${rollTarget.name}:rollOrb()`);
+  const github = await getOctokit();
+
+  try {
+    d(`roll triggered for  ${rollTarget.name}=${targetValue}`);
+
+    const filePath = '.circleci/config.yml';
+    const branchName = `roller/orb/${rollTarget.name}/${MAIN_BRANCH}`;
+    const shortRef = `heads/${branchName}`;
+    const ref = `refs/${shortRef}`;
+    const { owner, repo } = repository;
+
+    const response = await github.repos.getContent({
+      owner,
+      repo,
+      path: filePath,
+    });
+
+    if ('type' in response.data && 'content' in response.data && response.data.type == 'file') {
+      const content = Buffer.from(response.data.content, 'base64').toString();
+      const yamlData = yaml.parse(content);
+      let curr = yamlData[ORB_KEY];
+
+      // attempt to find the orb in .circleci/config.yml whos value includes `orbTarget.name`
+      const targetKey = Object.entries(curr as string).find(([_, value]) =>
+        value.startsWith(`${rollTarget.name}@`),
+      )?.[0];
+
+      if (targetKey === undefined) {
+        d(`Key for ${rollTarget.name} not found - skipping.`);
+        return;
+      }
+
+      // don't set the new value if the version is up to date
+      const previousVersion = curr[targetKey].split('@')[1];
+      if (targetValue === previousVersion) {
+        d(`No roll needed - ${rollTarget.name} is already at ${targetValue}`);
+        return;
+      }
+      curr[targetKey] = `${rollTarget.name}@${targetValue}`;
+
+      const newYamlData = yaml.stringify(yamlData);
```

Comment:
```suggestion
      const currentValue = curr[targetKey];
      const currentVersion = currentValue.split('@')[1];
      if (targetValue === currentVersion) {
        d(`No roll needed - ${rollTarget.name} is already at ${targetValue}`);
        return;
      }

      const currentValueRegex = new RegExp(`${rollTarget.name}@${currentVersion}`, 'g');
      let newYamlData: string;

      // If there is exactly one occurrence of the target string, we can get away with
      // doing a simple string replacement and avoid any potential formatting issues,
      // otherwise we need to stringify the full config which might change formatting
      if ((content.match(currentValueRegex) || []).length === 1) {
        newYamlData = content.replace(currentValueRegex, `${rollTarget.name}@${targetValue}`);
      } else {
        curr[targetKey] = `${rollTarget.name}@${targetValue}`;
        newYamlData = yaml.stringify(yamlData);
      }
```

I'm still concerned about PRs introducing formatting changes which will limit the usefulness of this feature. It's untested, but the above suggestion should mitigate that by doing a simple `String.replace` when it is safe to do so (which will leave formatting untouched), and falling back to the current `yaml.stringify` if it can't do a simple string replacement. We'll always get a PR out of it, but hopefully it takes the first path 99% of the time and won't muck with formatting.

Reply:
makes sense, applying the changes but keeping "previousValue" naming to match consistency with other rollers